File: lib/ai/field-resolver.ts
import "server-only";

import { createCustomField, updateCustomField } from "@/lib/db/queries/custom-fields";
import { customFieldTypeValues, type CustomFieldType, type FieldOption, type IterationOption } from "@/lib/db/schema";
import { fieldColorValues } from "@/lib/fields/colors";
import { slugify } from "@/lib/slug";

export type CustomFieldRow = {
  id: string;
  key: string;
  name: string;
  type: string;
  options: FieldOption[] | IterationOption[];
};

export type AiNewField = { key: string; name: string; type: string; options?: string[] };
export type AiFieldValue = { fieldKey: string; value: string };

function isValidType(type: string): type is CustomFieldType {
  return (customFieldTypeValues as readonly string[]).includes(type);
}

function isSelectType(type: CustomFieldType) {
  return type === "single_select" || type === "multi_select";
}

/** Applies the AI's proposed new fields + field values to the space's real
 * custom fields: creates any genuinely-missing field (matched by key,
 * case-insensitive, so a repeat paste never double-creates "Size") and
 * auto-grows select options for values that don't match an existing one —
 * this is the "AI task-creation agent can create necessary fields" behavior,
 * extended naturally to select options too. Mirrors resolveIdByName's role
 * for repo/milestone matching, just with DB writes since fields/options are
 * genuinely new data rather than a fixed set to match against.
 *
 * Returns a ready-to-persist `customFieldValues` map keyed by field id, plus
 * the (possibly grown) field list — including any field/option the AI just
 * created — so the caller can hand fresh field metadata back to the client
 * without a separate refetch. */
export async function resolveAiCustomFields(input: {
  spaceId: string;
  customFields: CustomFieldRow[];
  newFields?: AiNewField[];
  fieldValues?: AiFieldValue[];
}): Promise<{ customFieldValues: Record<string, unknown>; fields: CustomFieldRow[] }> {
  const fields = [...input.customFields];

  function findByKey(key: string) {
    return fields.find((f) => f.key.toLowerCase() === key.toLowerCase());
  }

  for (const proposal of (input.newFields ?? []).slice(0, 3)) {
    if (!proposal.key || !proposal.name || !isValidType(proposal.type)) continue;
    if (findByKey(proposal.key)) continue;

    const options: FieldOption[] = isSelectType(proposal.type)
      ? (proposal.options ?? []).slice(0, 10).map((label, i) => ({
          id: crypto.randomUUID(),
          name: label,
          color: fieldColorValues[i % fieldColorValues.length],
        }))
      : [];

    const created = await createCustomField({
      spaceId: input.spaceId,
      key: slugify(proposal.key) || slugify(proposal.name) || `field-${fields.length + 1}`,
      name: proposal.name,
      type: proposal.type,
      options,
    });
    fields.push({ id: created.id, key: created.key, name: created.name, type: created.type, options: created.options });
  }

  async function resolveOrCreateOptionId(field: CustomFieldRow, label: string): Promise<string | null> {
    const options = field.options as FieldOption[];
    const existing = options.find((o) => o.name.toLowerCase() === label.toLowerCase());
    if (existing) return existing.id;

    const newOption: FieldOption = {
      id: crypto.randomUUID(),
      name: label,
      color: fieldColorValues[options.length % fieldColorValues.length],
    };
    const nextOptions = [...options, newOption];
    await updateCustomField(field.id, { options: nextOptions });
    field.options = nextOptions; // keep the in-memory copy in sync for later lookups in this same request
    return newOption.id;
  }

  const result: Record<string, unknown> = {};

  for (const entry of (input.fieldValues ?? []).slice(0, 10)) {
    const field = findByKey(entry.fieldKey);
    const value = entry.value?.trim();
    if (!field || !value) continue;

    if (field.type === "text") {
      result[field.id] = value;
    } else if (field.type === "number") {
      const n = Number(value);
      if (!Number.isNaN(n)) result[field.id] = n;
    } else if (field.type === "date") {
      if (/^\d{4}-\d{2}-\d{2}$/.test(value)) result[field.id] = value;
    } else if (field.type === "single_select") {
      const id = await resolveOrCreateOptionId(field, value);
      if (id) result[field.id] = id;
    } else if (field.type === "multi_select") {
      const labels = value.split(",").map((v) => v.trim()).filter(Boolean);
      const ids: string[] = [];
      for (const label of labels) {
        const id = await resolveOrCreateOptionId(field, label);
        if (id) ids.push(id);
      }
      if (ids.length > 0) result[field.id] = ids;
    } else if (field.type === "iteration") {
      const opt = (field.options as IterationOption[]).find((o) => o.title.toLowerCase() === value.toLowerCase());
      if (opt) result[field.id] = opt.id;
    }
  }

  return { customFieldValues: result, fields };
}

File: Packages/ZenithAI/Sources/ZenithAI/FieldResolver.swift
import Foundation
import ZenithData

/// Port of `lib/ai/field-resolver.ts`. Applies the AI's proposed new fields
/// + field values to the space's real custom fields: creates any
/// genuinely-missing field (matched by key, case-insensitive, so a repeat
/// paste never double-creates "Size") and auto-grows select options for
/// values that don't match an existing one.
public enum FieldResolver {
    public struct ResolvedFields: Sendable {
        public let customFieldValues: [String: AnyCodableValue]
        public let fields: [CustomField]
    }

    public static func resolveAiCustomFields(
        _ db: ZenithDatabase, spaceId: UUID, customFields: [CustomField], newFields: [AINewField], fieldValues: [AIFieldValue]
    ) async throws -> ResolvedFields {
        var fields = customFields

        func findByKey(_ key: String) -> CustomField? {
            fields.first { $0.key.lowercased() == key.lowercased() }
        }

        for proposal in newFields.prefix(3) {
            guard !proposal.key.isEmpty, !proposal.name.isEmpty else { continue }
            guard findByKey(proposal.key) == nil else { continue }

            let isSelectType = proposal.type == .singleSelect || proposal.type == .multiSelect
            let options: FieldOptions
            if isSelectType {
                let labels = (proposal.options ?? []).prefix(10)
                let opts = labels.enumerated().map { index, label in
                    FieldOption(id: UUID().uuidString, name: label, color: FieldColors.values[index % FieldColors.values.count].rawValue)
                }
                options = .fields(opts)
            } else {
                options = .fields([])
            }

            // This resolver call site computes its own key preference
            // (proposed key, then name, then a numbered fallback) and talks
            // to `CustomFieldQueries` directly — deliberately bypassing
            // `CustomFieldActions.createCustomField`'s name-only key
            // derivation, matching the TS side's `resolveAiCustomFields`
            // (which imports from `lib/db/queries/custom-fields`, not the
            // actions layer, for exactly this reason).
            let slugifiedKey = Slug.slugify(proposal.key)
            let slugifiedName = Slug.slugify(proposal.name)
            let key = !slugifiedKey.isEmpty ? slugifiedKey : (!slugifiedName.isEmpty ? slugifiedName : "field-\(fields.count + 1)")
            let created = try await CustomFieldQueries.createCustomField(
                db, spaceId: spaceId, key: key, name: proposal.name, type: proposal.type, options: options
            )
            fields.append(created)
        }

        func resolveOrCreateOptionId(field: inout CustomField, label: String) async throws -> String? {
            var options = field.options.fieldOptions
            if let existing = options.first(where: { $0.name.lowercased() == label.lowercased() }) {
                return existing.id
            }
            let newOption = FieldOption(
                id: UUID().uuidString, name: label, color: FieldColors.values[options.count % FieldColors.values.count].rawValue
            )
            options.append(newOption)
            if let updated = try await CustomFieldQueries.updateCustomField(db, id: field.id, name: nil, options: .fields(options), position: nil) {
                field = updated
            }
            return newOption.id
        }

        var result: [String: AnyCodableValue] = [:]

        for entry in fieldValues.prefix(10) {
            guard let fieldIndex = fields.firstIndex(where: { $0.key.lowercased() == entry.fieldKey.lowercased() }) else { continue }
            let value = entry.value.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !value.isEmpty else { continue }
            var field = fields[fieldIndex]
            defer { fields[fieldIndex] = field }

            switch field.type {
            case .text:
                result[field.id.uuidString] = .string(value)
            case .number:
                if let number = Double(value) { result[field.id.uuidString] = .number(number) }
            case .date:
                if value.range(of: #"^\d{4}-\d{2}-\d{2}$"#, options: .regularExpression) != nil {
                    result[field.id.uuidString] = .string(value)
                }
            case .singleSelect:
                if let id = try await resolveOrCreateOptionId(field: &field, label: value) {
                    result[field.id.uuidString] = .string(id)
                }
            case .multiSelect:
                let labels = value.split(separator: ",").map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
                var ids: [String] = []
                for label in labels {
                    if let id = try await resolveOrCreateOptionId(field: &field, label: label) { ids.append(id) }
                }
                if !ids.isEmpty { result[field.id.uuidString] = .array(ids.map { .string($0) }) }
            case .iteration:
                if let option = field.options.iterationOptions.first(where: { $0.title.lowercased() == value.lowercased() }) {
                    result[field.id.uuidString] = .string(option.id)
                }
            }
        }

        return ResolvedFields(customFieldValues: result, fields: fields)
    }
}
